main.cpp

```diff
@@ -258,313 +207,113 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
 #include <iostream>
 #include <cmath>
 #include <functional>
+#include <memory>
 
-struct FloatType;
-struct IntType;
-struct DoubleType;
-
-struct FloatType
+template<typename MyTypeName> 
+struct Numeric
 {
-    FloatType& operator+=( float rhs );
-    FloatType& operator-=( float rhs );
-    FloatType& operator*=( float rhs );
-    FloatType& operator/=( float rhs );
-
-    operator float() const { return *value; }
+    using Type = MyTypeName;//treat as static?
 
-    FloatType& pow( float f );
-    FloatType& pow( const IntType& );
-    FloatType& pow( const FloatType& );
-    FloatType& pow( const DoubleType& );    
-
-    FloatType& apply(std::function<FloatType&(float&)> func); 
-    FloatType& apply(void(*)(float&));
-
-    explicit FloatType( float val ) : value( new float( val ) )
-    {        
-    }
-
-    ~FloatType()
+    Numeric& operator+=( const Type& rhs )
     {
-        delete value;
+        *value += rhs;
+        return *this;
     }
-
-    private:
-        float* value = nullptr;
-        FloatType& powInternal( const float exp );
-};
-
-struct DoubleType
-{
-    DoubleType& operator+=( double rhs );
-    DoubleType& operator-=( double rhs );
-    DoubleType& operator*=( double rhs );
-    DoubleType& operator/=( double rhs );
-
-    operator double() const {return *value;}
-
-    DoubleType& pow(double d);
-    DoubleType& pow(const IntType&);
-    DoubleType& pow(const FloatType&);
-    DoubleType& pow(const DoubleType&);  
-
-    DoubleType& apply(std::function<DoubleType&(double&)>);
-    DoubleType& apply(void(*)(double&));
-
-    explicit DoubleType( double val ) : value( new double(val) )
-    {        
+    Numeric& operator-=( const Type& rhs )
+    {
+        *value -= rhs;
+        return *this;
     }
-
-    ~DoubleType()
+    Numeric& operator*=( const Type& rhs )
     {
-        delete value;
+        *value *= rhs;
+        return *this;
     }
-
-    private:
-        double* value = nullptr;
-        DoubleType& powInternal( const double exp );
-};
-
-struct IntType
-{
-    IntType& operator+=( int rhs );
-    IntType& operator-=( int rhs );
-    IntType& operator*=( int rhs );
-    IntType& operator/=( int rhs );
-
-    IntType& pow( int i );
-    IntType& pow( const IntType& ) ;
-    IntType& pow( const FloatType& ) ;
-    IntType& pow( const DoubleType& ) ;
-
-    IntType& apply(std::function<IntType&(int&)> func);
-    IntType& apply(void(*)(int&));
-
-    operator int() const { return *value; }
-
-    explicit IntType( int val ) : value( new int( val ) ) { }
-    ~IntType()
+    template<typename T_>
+    Numeric& operator/=( const T_ rhs )
```

Comment:
Don't use `_` in type names
`T` is perfectly fine.  `T_` is unnecessary

- [ ] revise